File: host/cli/src/pus.rs
//! CCSDS space packets with PUS-C secondary headers (ECSS-E-ST-70-41C).

use crate::crc::crc16_ccitt;
use std::fmt;

pub const PRIMARY_HEADER: usize = 6;
pub const TC_SECONDARY_HEADER: usize = 5;
pub const TM_SECONDARY_HEADER: usize = 13;
pub const PUS_VERSION: u8 = 2;
/// Unix time of the mission epoch 2000-01-01T00:00:00Z.
pub const MISSION_EPOCH_UNIX: u64 = 946_684_800;

pub const ACK_ACCEPTANCE: u8 = 0x1;
pub const ACK_COMPLETION: u8 = 0x8;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Telecommand {
    pub apid: u16,
    pub sequence_count: u16,
    pub ack_flags: u8,
    pub service: u8,
    pub subtype: u8,
    pub source_id: u16,
    pub data: Vec<u8>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Telemetry {
    pub apid: u16,
    pub sequence_count: u16,
    pub service: u8,
    pub subtype: u8,
    pub message_counter: u16,
    pub destination_id: u16,
    /// CUC: seconds and 2^-16 fractions since the mission epoch.
    pub seconds: u32,
    pub fraction: u16,
    pub data: Vec<u8>,
}

impl Telemetry {
    /// Packet time as Unix milliseconds.
    pub fn unix_ms(&self) -> u64 {
        (MISSION_EPOCH_UNIX + u64::from(self.seconds)) * 1000
            + (u64::from(self.fraction) * 1000 + 32768) / 65536
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DecodeError {
    TooShort,
    BadVersion,
    LengthMismatch,
    WrongType,
    NoSecondaryHeader,
    BadPusVersion,
    Crc,
}

impl fmt::Display for DecodeError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let text = match self {
            DecodeError::TooShort => "too short",
            DecodeError::BadVersion => "bad packet version",
            DecodeError::LengthMismatch => "length mismatch",
            DecodeError::WrongType => "wrong packet type",
            DecodeError::NoSecondaryHeader => "no secondary header",
            DecodeError::BadPusVersion => "bad PUS version",
            DecodeError::Crc => "CRC error",
        };
        f.write_str(text)
    }
}

impl std::error::Error for DecodeError {}

fn primary_header(tc: bool, apid: u16, seq: u16, data_len: usize) -> [u8; PRIMARY_HEADER] {
    let word0 = (u16::from(tc) << 12) | 0x0800 | (apid & 0x7FF);
    let word1 = 0xC000 | (seq & 0x3FFF);
    let len = (data_len - 1) as u16;
    let mut h = [0u8; PRIMARY_HEADER];
    h[0..2].copy_from_slice(&word0.to_be_bytes());
    h[2..4].copy_from_slice(&word1.to_be_bytes());
    h[4..6].copy_from_slice(&len.to_be_bytes());
    h
}

/// Encodes a telecommand, CRC included.
pub fn encode_tc(tc: &Telecommand) -> Vec<u8> {
    let data_len = TC_SECONDARY_HEADER + tc.data.len() + 2;
    let mut p = Vec::with_capacity(PRIMARY_HEADER + data_len);
    p.extend_from_slice(&primary_header(true, tc.apid, tc.sequence_count, data_len));
    p.push((PUS_VERSION << 4) | (tc.ack_flags & 0x0F));
    p.push(tc.service);
    p.push(tc.subtype);
    p.extend_from_slice(&tc.source_id.to_be_bytes());
    p.extend_from_slice(&tc.data);
    let crc = crc16_ccitt(&p);
    p.extend_from_slice(&crc.to_be_bytes());
    p
}

/// Encodes a telemetry packet (used by the tests and simulators).
pub fn encode_tm(tm: &Telemetry) -> Vec<u8> {
    let data_len = TM_SECONDARY_HEADER + tm.data.len() + 2;
    let mut p = Vec::with_capacity(PRIMARY_HEADER + data_len);
    p.extend_from_slice(&primary_header(false, tm.apid, tm.sequence_count, data_len));
    p.push(PUS_VERSION << 4);
    p.push(tm.service);
    p.push(tm.subtype);
    p.extend_from_slice(&tm.message_counter.to_be_bytes());
    p.extend_from_slice(&tm.destination_id.to_be_bytes());
    p.extend_from_slice(&tm.seconds.to_be_bytes());
    p.extend_from_slice(&tm.fraction.to_be_bytes());
    p.extend_from_slice(&tm.data);
    let crc = crc16_ccitt(&p);
    p.extend_from_slice(&crc.to_be_bytes());
    p
}

fn be16(p: &[u8], off: usize) -> u16 {
    u16::from_be_bytes([p[off], p[off + 1]])
}

/// Decodes and checks a telemetry packet.
pub fn decode_tm(p: &[u8]) -> Result<Telemetry, DecodeError> {
    if p.len() < PRIMARY_HEADER {
        return Err(DecodeError::TooShort);
    }
    if p[0] >> 5 != 0 {
        return Err(DecodeError::BadVersion);
    }
    let word0 = be16(p, 0);
    if PRIMARY_HEADER + usize::from(be16(p, 4)) + 1 != p.len() {
        return Err(DecodeError::LengthMismatch);
    }
    if word0 & 0x1000 != 0 {
        return Err(DecodeError::WrongType);
    }
    if word0 & 0x0800 == 0 {
        return Err(DecodeError::NoSecondaryHeader);
    }
    if p.len() < PRIMARY_HEADER + TM_SECONDARY_HEADER + 2 {
        return Err(DecodeError::TooShort);
    }
    if p[PRIMARY_HEADER] >> 4 != PUS_VERSION {
        return Err(DecodeError::BadPusVersion);
    }
    if crc16_ccitt(p) != 0 {
        return Err(DecodeError::Crc);
    }
    let s = PRIMARY_HEADER;
    Ok(Telemetry {
        apid: word0 & 0x7FF,
        sequence_count: be16(p, 2) & 0x3FFF,
        service: p[s + 1],
        subtype: p[s + 2],
        message_counter: be16(p, s + 3),
        destination_id: be16(p, s + 5),
        seconds: u32::from_be_bytes([p[s + 7], p[s + 8], p[s + 9], p[s + 10]]),
        fraction: be16(p, s + 11),
        data: p[s + TM_SECONDARY_HEADER..p.len() - 2].to_vec(),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    /// TC[17,1] to APID 0x010, sequence 0, source 1, acknowledgement 0x9, as produced by the
    /// flight library (satlink::pus::Encode).
    const PING: [u8; 13] = [
        0x18, 0x10, 0xC0, 0x00, 0x00, 0x06, 0x29, 0x11, 0x01, 0x00, 0x01, 0, 0,
    ];

    #[test]
    fn telecommand_layout() {
        let tc = Telecommand {
            apid: 0x010,
            sequence_count: 0,
            ack_flags: ACK_ACCEPTANCE | ACK_COMPLETION,
            service: 17,
            subtype: 1,
            source_id: 1,
            data: vec![],
        };
        let p = encode_tc(&tc);
        assert_eq!(&p[..11], &PING[..11]);
        assert_eq!(crc16_ccitt(&p), 0);
    }

    #[test]
    fn telemetry_round_trip() {
        let tm = Telemetry {
            apid: 0x010,
            sequence_count: 0x3FFF,
            service: 3,
            subtype: 25,
            message_counter: 7,
            destination_id: 1,
            seconds: 815_000_000,
            fraction: 0x8000,
            data: vec![1, 2, 3],
        };
        let p = encode_tm(&tm);
        assert_eq!(p.len(), 6 + 13 + 3 + 2);
        assert_eq!(decode_tm(&p), Ok(tm.clone()));
        assert_eq!(tm.unix_ms(), (946_684_800 + 815_000_000) * 1000 + 500);

        let mut bad = p.clone();
        bad[20] ^= 1;
        assert_eq!(decode_tm(&bad), Err(DecodeError::Crc));
        assert_eq!(decode_tm(&p[..10]), Err(DecodeError::LengthMismatch));
        assert_eq!(decode_tm(&p[..3]), Err(DecodeError::TooShort));
        let mut tc = p.clone();
        tc[0] |= 0x10;
        assert_eq!(decode_tm(&tc), Err(DecodeError::WrongType));
        let mut version = p.clone();
        version[0] |= 0x20;
        assert_eq!(decode_tm(&version), Err(DecodeError::BadVersion));
        let mut pus = p;
        pus[6] = 0x10;
        assert_eq!(decode_tm(&pus), Err(DecodeError::BadPusVersion));
        assert_eq!(DecodeError::Crc.to_string(), "CRC error");
    }
}

File: host/python/satlink/pus.py
"""CCSDS space packets (133.0-B-2) with PUS-C secondary headers (ECSS-E-ST-70-41C).

Same layout as libs/pus: TC data field header of 5 bytes, TM data field header of 13 bytes with
a CUC time (4 + 2 bytes) since 2000-01-01T00:00:00Z, CRC-16-CCITT at the end of every packet.

@implements SRS-GS-004
"""
from __future__ import annotations

import struct
from dataclasses import dataclass, field

PUS_VERSION = 2
MISSION_EPOCH_UNIX = 946_684_800
ACK_ACCEPTANCE = 0x1
ACK_COMPLETION = 0x8


def crc16_ccitt(data: bytes, crc: int = 0xFFFF) -> int:
    """CRC-16-CCITT, polynomial 0x1021, initial value 0xFFFF."""
    for byte in data:
        crc ^= byte << 8
        for _ in range(8):
            crc = ((crc << 1) ^ 0x1021) if crc & 0x8000 else (crc << 1)
            crc &= 0xFFFF
    return crc


class DecodeError(ValueError):
    """The bytes are not a valid PUS telemetry packet."""


@dataclass
class Telecommand:
    apid: int
    sequence_count: int
    service: int
    subtype: int
    data: bytes = b""
    ack_flags: int = ACK_ACCEPTANCE | ACK_COMPLETION
    source_id: int = 1

    def encode(self) -> bytes:
        body = bytes([(PUS_VERSION << 4) | (self.ack_flags & 0xF), self.service, self.subtype])
        body += struct.pack(">H", self.source_id) + bytes(self.data)
        return _finish(True, self.apid, self.sequence_count, body)


@dataclass
class Telemetry:
    apid: int
    sequence_count: int
    service: int
    subtype: int
    message_counter: int = 0
    destination_id: int = 1
    seconds: int = 0
    fraction: int = 0
    data: bytes = field(default=b"")

    @property
    def unix_time(self) -> float:
        return MISSION_EPOCH_UNIX + self.seconds + self.fraction / 65536.0

    def encode(self) -> bytes:
        body = bytes([PUS_VERSION << 4, self.service, self.subtype])
        body += struct.pack(">HHIH", self.message_counter, self.destination_id, self.seconds,
                            self.fraction) + bytes(self.data)
        return _finish(False, self.apid, self.sequence_count, body)

    @classmethod
    def decode(cls, packet: bytes) -> "Telemetry":
        if len(packet) < 6:
            raise DecodeError("too short")
        word0, word1, length = struct.unpack_from(">HHH", packet)
        if packet[0] >> 5:
            raise DecodeError("bad packet version")
        if 6 + length + 1 != len(packet):
            raise DecodeError("length mismatch")
        if word0 & 0x1000:
            raise DecodeError("not a telemetry packet")
        if not word0 & 0x0800:
            raise DecodeError("no secondary header")
        if len(packet) < 6 + 13 + 2:
            raise DecodeError("too short")
        if packet[6] >> 4 != PUS_VERSION:
            raise DecodeError("bad PUS version")
        if crc16_ccitt(packet) != 0:
            raise DecodeError("CRC error")
        service, subtype = packet[7], packet[8]
        counter, dest, seconds, fraction = struct.unpack_from(">HHIH", packet, 9)
        return cls(word0 & 0x7FF, word1 & 0x3FFF, service, subtype, counter, dest, seconds,
                   fraction, bytes(packet[19:-2]))


def _finish(tc: bool, apid: int, seq: int, body: bytes) -> bytes:
    data_len = len(body) + 2
    header = struct.pack(">HHH", (int(tc) << 12) | 0x0800 | (apid & 0x7FF),
                         0xC000 | (seq & 0x3FFF), data_len - 1)
    packet = header + body
    return packet + struct.pack(">H", crc16_ccitt(packet))
